### ISSUE TITLE
SSL

### ISSUE THREAD

**Issue by DanGrayson** [Macaulay2/Macaulay2-web-site#8]
We should switch to SSL-only service for increased security for our users.

**Comment by DanGrayson** [comment]
We already are on https, by math department policy, so the only non-SSL site is our mirror.

**Comment by mahrud** [comment]
Two ways to have https for the mirror:
- Cloudflare provides free tls certificates (near zero setup)
- https://letsencrypt.org provides is a free certificate authority with automatic tooling for updating the certificate.

**Comment by mahrud** [comment]
Why is this under discussion? Currently https://macaulay2.com/ doesn't work, and two potential solutions was provided.

**Comment by mahrud** [comment]
@DanGrayson None of the links in your emails today work for me, as they all use https. The fix is nearly trivial! Do you need help setting it up?
```
[mahrud@noether ~]$ curl https://macaulay2.com
curl: (28) Failed to connect to macaulay2.com port 443: Connection timed out
```

**Comment by DanGrayson** [comment]
Just try this instead:
```
$ curl macaulay2.com
<html><body>You are now redirected to<br><a href="https://www.math.illinois.edu/Macaulay2/">https://www.math.illinois.edu/Macaulay2/</a></body></html>
```

**Comment by mahrud** [comment]
This is trivial to fix and would cost $0, if anybody cared.

@mikestillman is it okay with you that the links in the emails don't work, and the offered solution is to "remove the 's'"?

**Comment by DanGrayson** [comment]
I finally figured out which "emails" you're referring to.  I'll see what I can do.

**Comment by DanGrayson** [comment]
Joker.com doesn't seem to offer https url forwarding (see https://joker.com/faq/content/52/111/en/what-is-url-forwarding-web-forwarding.html ).  I'll send out a correction.

**Comment by mahrud** [comment]
...

You can switch to using Cloudflare as the name service, which only charges you the amount that they need to pay Joker for registrar fees. Here is the doc: https://developers.cloudflare.com/registrar/

Then proxy the traffic through them. There, you can enable TLS for free. They will provide the certificate and take care of renewals. This also reduces load on the server.

I have worked for the company and know how their website works. Would you like me to do it for you?

All of this costs nothing extra. You can also do it without Cloudflare through Let's Encrypt, but the config is more involved.

The fact that this issue has been open for years is really silly.

**Comment by DanGrayson** [comment]
Yes, but that doesn't address the issue at all, which is that our mirror (on one of my virtual machines) serves up http.


Re: "The fact that this issue has been open for years is really silly."

It's been 1 year.

**Comment by mahrud** [comment]
> Yes, but that doesn't address the issue at all, which is that our mirror (on one of my virtual machines) serves up http.

Dan, I don't think you understand what Cloudflare does. Cloudflare sits between your server and the internet. **It serves https to the world**, and queries your server for requests that are not cached. If a request is cached, it'll show the result from it's servers across the globe. This will significantly reduce load on your server, **and** provide https for users.

Would you like to call and do it together?

**Comment by mahrud** [comment]
To reiterate: your server doesn't need to serve https, http is enough. Cloudflare's servers have https servers running and serving the content that they get through http from your server.

**Comment by DanGrayson** [comment]
Oh!  Then it's worth trying.  

Either that, or we could just get rid of the mirror.  Our load is rather light, I'm sure.

**Comment by DanGrayson** [comment]
On second thought, if the data passes insecurely from cloudflare to our server, then the connection is insecure, so there doesn't seem to be any point in doing it.

**Comment by mahrud** [comment]
Attacks don't happen between your server server and Cloudflare (if they do,
you need to switch your internet provider!). They happen near the user.

Regardless, the problem right now is that https://macaulay2.com doesn't
work. Can you get https to work on your server? If not then this is the
easiest solution.

**Comment by DanGrayson** [comment]
Re: "the problem right now is that https://macaulay2.com doesn't work"

That's no longer a problem -- I sent out a correction giving the correct URLs.

**Comment by mahrud** [comment]
That's a non-answer. If you can't do it, or don't want to do it, then just
close this issue.

**Comment by mahrud** [comment]
At least try it for a while, if you don't notice a significant raise in how fast the website loads, then revert it.

**Comment by DanGrayson** [comment]
> At least try it for a while, if you don't notice a significant raise in how fast the website loads, then revert it.

I won't notice any speed increase, as I'm 3 miles from macaulay2.com.

**Comment by mahrud** [comment]
You will. I guarantee it. Just give it a try.

**Comment by dimpase** [comment]
Cloudflare's sales pitch strikes be as biased (oh well, surprise).

Setting up letsencrypt does not require changing internet providers, and does not tie you up to yet another dominant Internet player. I'd do letsencrypt.

**Comment by DanGrayson** [comment]
Is "letsencrypt"'s method like cloudflare's method, in that the portion of the communication channel between letsencrypt and our server is unsecure?

**Comment by dimpase** [comment]
No. letsencrypt issues you a normal SSL certificate, which needs to be updated every month or so.
Clients to do this renewal are available for every major OS.

**Comment by dimpase** [comment]
[letsencrypt](https://en.wikipedia.org/wiki/Let%27s_Encrypt) is a true open-source solution, too.

**Comment by DanGrayson** [comment]
Ah, so letsencrypt is simply a way to get a certificate for free, and then we use it to set up SSL on our server.  That sounds better than cloudflare, in that we retain control of our certificate and there is no insecure portion of the communication channel.

How do the certificates issued by letsencrypt get trusted by all the browsers out there?  Is letsencrypt trusted by a higher authority?

**Comment by mahrud** [comment]
They are a trusted root authority: https://letsencrypt.org/certificates

That's also a fine solution. I use Let's Encrypt myself, plus Cloudflare's service for it's speedup and caching. In particular, there's already an issue (#10) to use cloud fronting, so I don't understand @dimpase's objection.

**Comment by dimpase** [comment]
> In particular, there's already an issue (#10) to use cloud fronting, so I don't understand @dimpase's objection.

I don't understand why cloudflare has a free tier. Probably it's a way to collect fine-grained browsing stats - and sell it to
various adware dealers.

**Comment by DanGrayson** [comment]
> In particular, there's already an issue (#10) to use cloud fronting, so I don't understand @dimpase's objection.

To be fair to Dima, the existence of an issue suggesting to consider whether to do something doesn't mean a decision has been made to do it.

**Comment by mahrud** [comment]
> I don't understand why cloudflare has a free tier. Probably it's a way to collect fine-grained browsing stats - and sell it to
various adware dealers.

That's a reasonably skeptical take on a tech company these days, but I can tell you for sure that they're not in the business of selling data 🙂 They have a free tier for the same reason that antivirus companies have a free service: to find out about attacks out in the wild early. They make enough money from paid users.

Also, does SageMath sell data? :P

ps: I'm not getting paid by them.

**Comment by mahrud** [comment]
> To be fair to Dima, the existence of an issue suggesting to consider whether to do something doesn't mean a decision has been made to do it.

Sure, but there's no virtue in remaining "open source" here.

**Comment by DanGrayson** [comment]
I'm leaning toward setting up a new virtual machine online, calling it www1.macaulay2.com temporarily (or just upgrading www2 from Ubuntu 18.04 to Ubuntu 20.04), using letsencrypt to get a certificate, using apache as the server, and making it the primary site eventually.  After all, I won't be around forever, and I haven't heard anyone say that my math department wants to be involved in our activity in perpetuity.

No one has ever complained about download speed, and cloudflare's solution seems insecure to me, so I don't see the point of pursuing that.

**Comment by DanGrayson** [comment]
PS:  Could we forget about virtual machines and do everything through github?  If so, how would that actually work?

**Comment by dimpase** [comment]
you can definitely host a website on GitHub Pages, one that would be quite sufficient for M2 purposes, and
do binary releases using GitHub's releases (one can "edit" the release and upload whatever "binary assets" you want - there are some limits on file sizes, but IMHO M2 is quite OK in this regard).

**Comment by DanGrayson** [comment]
You're right - I was able to upload all my binary release files to the github release.

**Comment by DanGrayson** [comment]
Is there a limit for github pages total size?

**Comment by dimpase** [comment]
[1 Gb for GH Pages](https://docs.github.com/en/free-pro-team@latest/github/working-with-github-pages/about-github-pages#usage-limits)

Needless to say, it's the website only, source code and releases of the project don't count.

**Comment by mahrud** [comment]
Also they're static, so the "random paper referring to Macaulay2" widget wouldn't work.

**Comment by DanGrayson** [comment]
Hmm, why doesn't this work now?  https://macaulay2.github.io/foo.html

https://github.com/Macaulay2/Macaulay2.github.io

**Comment by mahrud** [comment]
I don't know what you mean. Typically github pages are created using markdown and populated using Jekyll, rather than uploading html files. This was partially the goal of the `markdown` command that I wrote, so you could make package documentation like this: https://mahrud.github.io/m2doc/packages/FGLM/ (this is a work in progress)

**Comment by DanGrayson** [comment]
Oh, now it started working.  I guess there was a sizable delay.

**Comment by dimpase** [comment]
>  "random paper referring to Macaulay2" widget wouldn't work.

IMHO Pages do allow embedded javascript (perhaps curated somehow). So YMMV.

**Comment by DanGrayson** [comment]
> [1 Gb for GH Pages](https://docs.github.com/en/free-pro-team@latest/github/working-with-github-pages/about-github-pages#usage-limits)
> 
> Needless to say, it's the website only, source code and releases of the project don't count.

I think the repository might be 0.5gb if we include the documentation for 2 versions.  But then after a new version is
released, we need a way to prune the older documentation from the repository, so we don't exceed our limit.

**Comment by dimpase** [comment]
one can split the site into parts, e.g. create a dedicated GH Pages site for packages, a dedicated site for docs, etc.

**Comment by DanGrayson** [comment]
Splitting it up doesn't reduce the total size.

**Comment by DanGrayson** [comment]
Here's a procedure for removing all the history from a repository, but keeping the files:
```sh
    git checkout --orphan new-master
    git branch -D master
    git branch --move master
    git reflog expire --all --expire=now
    git gc --prune=now --aggressive
    git add -A
    git commit -m 'add all the files'
```

**Comment by dimpase** [comment]
> Splitting it up doesn't reduce the total size.

You can create, say, github.com/Macaulay2-packages
and associate a separate GH Pages with it, for packages info.
This won't count into your 1Gb quota for github.com/Macaulay2

**Comment by DanGrayson** [comment]
Well, eventually that will fill up, too, if we keep adding documentation html to it.

**Comment by dimpase** [comment]
1Gb of html documents seems to me  a truly huge amount of docs.

**Comment by DanGrayson** [comment]
I've just checked - the html documentation for 1.17 is 125mb.  So 8 of those will fill it.  4 years.

**Comment by dimpase** [comment]
I wouldn't worry about the docs for more than 2 releases at the same time.

**Comment by dimpase** [comment]
By the way, Sagemath's html docs are 23Mb, it's strange that M2's docs, smaller by pagecount, IMHO, take so much more space.

**Comment by DanGrayson** [comment]
Actually, if I remove the example output files, it's just 93mb, with 9000 files.  So 10K per html file on average.  We have lot of html that's automatically generated -- that could explain some of the size.

**Comment by mahrud** [comment]
Alternatively, we can use the markdown version of the documentation. With markdown, the template is only written in a single file, so we save 9000*[the size of our html header section]. This is a benefit of jekyll. See http://jekyllbootstrap.com/lessons/jekyll-introduction.html.

**Comment by DanGrayson** [comment]
> Alternatively, we can use the markdown version of the documentation. With markdown, the template is only written in a single file, so we save 9000*[the size of our html header section]. This is a benefit of jekyll. See http://jekyllbootstrap.com/lessons/jekyll-introduction.html.

I don't follow the logic here, since the html files have to be static and have to be generated, but as an experiment, it might be worthwhile generating all the html files your way, to see if it's less than 93mb, and to show us the result.

**Comment by dimpase** [comment]
I noticed that many html files in the docs consist for ~ 50% of rather long
links to top/next/previous docs nodes.

**Comment by DanGrayson** [comment]
Yes, here are some examples:
```html
          <li>
<span><tt>&quot;addHook(...,Strategy=>...)&quot;</tt> -- see <span><a title="add a hook function to an object for later processing" href="../../Macaulay2Doc/html/_add__Hook.html">addHook</a> -- add a hook function to an object for later processing</span></span>          </li>
          <li>
<span><tt>&quot;annihilator(...,Strategy=>...)&quot;</tt> -- see <span><a title="the annihilator ideal" href="../../Saturation/html/_annihilator.html">annihilator</a> -- the annihilator ideal</span></span>          </li>
          <li>
<span><tt>&quot;associatedPrimes(...,Strategy=>...)&quot;</tt> -- see <span><a title="find associated primes" href="../../PrimaryDecomposition/html/_associated__Primes.html">associatedPrimes</a> -- find associated primes</span></span>          </li>
          <li>
<span><tt>&quot;mingens(...,Strategy=>...)&quot;</tt> -- see <span><a title="a Strategy option value" href="../../Macaulay2Doc/html/___Complement.html">Complement</a> -- a Strategy option value</span></span>          </li>
```

**Comment by mahrud** [comment]
> I don't follow the logic here, since the html files have to be static and have to be generated, but as an experiment, it might be worthwhile generating all the html files your way, to see if it's less than 93mb, and to show us the result.

With jekyll, github generates the html files, not us. The repository contains only markdown files plus a few template files written in html.

**Comment by mahrud** [comment]
Regardless, this is a bigger project that probably won't happen for some time. We've collectively provided multiple ways to get TLS to work, all of which are improvements over the current situation.

**Comment by DanGrayson** [comment]
Oh, okay.
